mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "adjoint"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+template <class T> T isInstanceOf(Operation &op)
+{
+    if (op.getName().getStringRef() == T::getOperationName())
+        return cast<T>(op);
+    else
+        return nullptr;
+}
+
+/// Copy the region of the adjoint operation `op` to the POI specified by the `rewriter`. Build and
+/// return the value mapping `bvm`.
+Value copyAdjointVerbatim(AdjointOp op, PatternRewriter &rewriter, IRMapping &bvm)
+{
+    Block &b = op.getRegion().front();
+    for (auto i = b.begin(); i != b.end(); i++) {
+        if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+            assert(++i == b.end() &&
+                   "quantum.yield must be the last operation of an adjoint block");
+            return bvm.lookupOrDefault(yield->getOperand(0));
+        }
+        else {
+            rewriter.insert(i->clone(bvm));
+        }
+    }
+    assert(false && "quantum.yield must present in the adjoint region");
+}
+
+struct AdjointSingleOpRewritePattern : public mlir::OpRewritePattern<AdjointOp> {
+    using mlir::OpRewritePattern<AdjointOp>::OpRewritePattern;
+
+    /// In essence, we build a map from values mentiond in the source data flow to the values of the
+    /// program where quantum control flow is reversed. Most of the time, there is a 1-to-1
+    /// correspondence with a notable exception caused by `insert`/`extract` API asymetry.
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
```

Comment:
This can also be made explicit in the op definition (see https://mlir.llvm.org/docs/Traits/#single-block-region)

Reply:
Sorry, this one might actually be better: https://mlir.llvm.org/docs/Traits/#single-block-with-implicit-terminator
because it also enforces the specific terminator type.

Reply:
After some discussion I have applied the trait but at the same time continued to emit `quantum.YieldOp` explicitly.